### ISSUE TITLE
Dropping the develop Branch

### ISSUE THREAD

**Issue by egeldenhuys** [TripleParity/docks#23]
EDIT: The proposed workflow is similar to the [GitHub Flow](https://guides.github.com/introduction/flow/)

After reviewing or protected branch policy I have seen that the develop branch is redundant. Since this observation goes against our workflow and my personal workflow I will discuss the situation in detail.

Firstly I will reiterate our current workflow. I will then show that develop and master are equivalent when using tags. I will then refute my own counter argument. Finally I will discuss the advantages of merging features into the master branch.

### Current Workflow

Feature branches are created from develop. The feature is then implemented and a pull request (PR) is made to merge into develop. At this point the code in develop is peer reviewed and tested. Travis CI performs a "pr" tests where it tests the code as if it has already been merged with develop. Once the reviewers are happy with the code it is merged into develop.

Once we or I feel develop is stable and contains enough code to release, it is merged into master and tagged with a version. The only changes to master are by merging develop or hotfix branches.

### Develop and Master are Equivalent
Currently we treat master as "production ready" and "stable". Since only develop is merged into master to signal a release master performs the same functionality as a tag. Then we also tag the merge commit in master. A tag is a reference to a single commit, such as "0.0.2".

I propose instead of merging into develop, we merge into master. When want to signal to the public that they can upgrade to a new version, we tag a commit in master with a version. Clearly how one interprets master needs to change since master will contain unreleased work. If you want to deploy a stable version of the product, you need to checkout a tagged version. This is the only mindset that has to change - master will contain unreleased work.

### Counter Arguments
I will now address some thoughts I had against this proposal.

#### But develop is used for integrating and testing features. By merging into master is it more likely to introduce uncaught bugs into releases

While it is true that we often discover bugs while developing features using the develop codebase this is no different from discovering bugs while using the master codebase. This means that master will contain code that has recently been written and perhaps contain uncaught bugs. However if develop can reach a point where it is considered "stable" and ready to be released, then so can master. At that point we tag it with a version.

When using tags as stable versions, the unreleased code in master will not affect production systems.

Furthermore since we follow a continuous integration cycle, code will be tested on multiple levels before being merged:
1. Unit testing docks-api
2. Integration testing docks-api
3. Unit testing docks-ui
4. Integration testing docks-ui with docks-api
5. Peer reviews during PR

#### If we work with master it is more likely we will mess up the branch

The branch is protected from forced pushes and deletions. Furthermore pull request reviews are required before merging.

#### Master will always reflect a working and deployable prototype

This is a requirement by the COS 301 course. No longer will there be questions relating to "who contributed" as contributions will display immediately after a feature is merged.

Since the code has been extensively tested, the unreleased version should be production ready.

#### GitHub will work in our favour

Currently referenced issues have to be closed manually after merging into develop. However GitHub will do this automatically if the PR or commit is merged into the default branch, i.e master.

#### Simpler workflow
If a README in master has to be updated with important information a hotfix branch has to be created from master and merged back into master. With our current workflow it also has to be merged into develop.

To get code from develop into master either a pull request has to be made
and reviewed or it has to be merged manually. This step would not be required if we only use master.

#### More respect for Pull Request reviews

In my mind approving and merging code into develop is safe. develop cannot break a production system as they should not be deploying from develop. However at any time develop can be merged into master and deployed to a production system. This is a false sense of security in my mind.

#### New feature

A branch is made from the HEAD of master. The feature is implemented and a PR is made to merge into master

#### Hotfix

After making some commits to master we discover the last released version has a critical bug. So we create a hotfix branch from the tag, implement the fix and tag a patch version in that branch. We then merge the hotfix branch into master to benefit future releases.

#### Deploying the latest development version

The master branch is checked out, and the Docker image is built.

#### Deploying the latest stable version

The latest tag is checkout out, and the Docker image is built.

### Summary

- The master and develop branches are equivalent if you accept that master contains unreleased code and releases are tagged.
- Since the code is thoroughly tested, it should work in production.
- Merging into master allows GitHub to automatically close issues for us.
- Contributions will immediately show once a feature is merged.
- The workflow is simpler.

### Request For Comments

- If there are any scenarios that will not work with the proposed workflow please raise them so we can reason about them
- If anything is unclear please ask
- If you do not agree please state why
- If you agree please let us know
